README

avmount
=======

avmount - Mount MediaServers content as a Linux filesystem.
Created 23/04/2005, Rémi Turboult <r3mi@users.sourceforge.net>


avmount is a UPnP AV client based on djmount. Originally created
by Rémi Turboult <r3mi@users.sourceforge.net> on 23/04/2005. It mounts as a
Linux filesystem (using FUSE) the media content of compatible UPnP AV devices
(see "Known Compatible Devices" paragraph below).

avmount discovers automatically all UPnP AV Media Servers on the network,
and make the content available in a directory tree. All shared files 
(e.g. Audio or Video files) are directly visible and can be played using 
your favorite media player.


avmount is free software, licensed under the terms of the GNU General Public
License (GNU GPL : see "COPYING" file) : see details in "Copyright and 
Licenses" paragraph below.


Example 1 : browsing
--------------------

(the actual files listed depend on your AV files collection ! Also, the 
directory structure is different for each kind of Media Server)

$ avmount /mnt/av

$ cd /mnt/av

$ ls -l
dr-xr-xr-x  3 root root 512 jan  1  2000 SomeMediaServerDeviceName/
dr-xr-xr-x  3 root root 512 jan  1  2000 AnotherMediaServerDeviceName/

$ cd SomeMediaServerDeviceName

$ ls -l 
dr-xr-xr-x  14 root root 512 jan  1  2000 Music/
dr-xr-xr-x   5 root root 512 jan  1  2000 Pictures/
dr-xr-xr-x   4 root root 512 jan  1  2000 Videos/

$ ls -l Music/Artists/Tinariwen/
-r--r--r--  1 root root 5407087 jan  1  2000 Chatma.mp3
-r--r--r--  1 root root 3751143 jan  1  2000 Chet Boghassa.mp3

$ xmms Music/Artists/Tinariwen/

$ fusermount -u /mnt/av


Note: the FUSE  kernel module shall be loaded before running avmount 
(otherwise you will get the error: 
"fusermount: fuse device not found, try 'modprobe fuse' first").


Example 2 : searching
---------------------

(again, the actual content listed depends on your AV files collection !)

$ avmount /mnt/av

$ cd /mnt/av

$ ls -l
dr-xr-xr-x  3 root root 512 jan  1  2000 SomeMediaServerSupportingSearch/
dr-xr-xr-x  3 root root 512 jan  1  2000 AnotherMediaServerDeviceName/

$ cd SomeMediaServerSupportingSearch

$ ls -l 
dr-xr-xr-x  14 root root 512 jan  1  2000 Music/
dr-xr-xr-x   5 root root 512 jan  1  2000 Pictures/
dr-xr-xr-x   4 root root 512 jan  1  2000 Videos/
dr-xr-xr-x   2 root root 512 Jan  1  2000 _search/

$ ls -l Music/_search
-r--r--r--  1 root root   126 Jan  1  2000 search_capabilities
-r--r--r--  1 root root 12640 Jan  1  2000 search_help.txt

$ ls -l Music/_search/brel
-r--r--r--  1 root root 3151929 jan  1  2000 Amsterdam.mp3
dr-xr-xr-x  4 root root     512 jan  1  2000 Jacques Brel/
-r--r--r--  1 root root 3246781 jan  1  2000 Ne me quitte pas.mp3
dr-xr-xr-x  2 root root     512 Jan  1  2000 _and/
dr-xr-xr-x  2 root root     512 Jan  1  2000 _or/

$ xmms Music/_search/brel/Amsterdam.mp3

$ fusermount -u /mnt/av


Note: full help for searching is available in the "search_help.txt" file,
included in the avmount package. This file is also available at run-time 
in the "_search" directory, if the connected UPnP device supports searching.


Building and Installation
-------------------------

avmount has been built and tested on various Linux distro (MandrakeLinux 10.1,
GeeXboX, Debian sarge or sid, Ubuntu 5.10, Mandriva 2006.0 ...), which include
various Linux kernels (e.g. 2.6.14), gcc (e.g. 3.4.1, 4.0.3) and libc
(e.g. glibc-2.3.5, uClibc 0.9.28).
Different versions of FUSE are also supported (e.g. 2.2.1, 2.4, 2.5)


A) Basic build instruction:

1) Prerequisites for compilation:

   * the FUSE package (2.9.4) : http://fuse.sourceforge.net/
   * the Portable SDK for UPnP Devices (2.6.18) : http://pupnp.sourceforge.net/
   * the cURL package (7.45) : https://curl.haxx.se/
   * the talloc library (2.1.5) : https://talloc.samba.org/

2) Use 'configure' and 'make' in the top directory to compile.
   Example :
	./configure
	make

3) if you want to install avmount on your system, run :
	make install

   or to strip the resulting executable :
	make install-strip


B) Advanced build instruction:

"configure" as a few options to customise avmount: see --help option.
In particular,
	
  --disable-debug             extra debugging code [default=enabled]
  --with-fuse-prefix=DIR      search for fuse in DIR/include and DIR/lib
  --with-libiconv-prefix=DIR  search for libiconv in DIR/include and DIR/lib
  --with-talloc-prefix=DIR    search for talloc in DIR/include and DIR/lib
  --with-libupnp-prefix=DIR   search for libupnp in DIR/include and DIR/lib

configure with also checks the following environment variables in
case there is a custom installation (see also configure --help):
	FUSE_CFLAGS, FUSE_LIBS, 
	TALLOC_CFLAGS, TALLOC_LIBS, 
	LIBUPNP_CFLAGS, LIBUPNP_LIBS.

Examples:
	./configure --with-external-talloc --with-talloc-prefix=/opt/local
	TALLOC_CFLAGS='-I/opt/include' ./configure --with-external-talloc

Usage 
-----

1) as root : load the FUSE kernel module

	modprobe fuse

2) To run, simply use the 'avmount' executable
   (no extra file is needed).
   The FUSE mountpoint is a required argument. For example:

	avmount /mnt/upnp

3) when finished, unmount the mountpoint (see details in FUSE documentation):

	fusermount -u /mnt/upnp

4) Options :
   see "avmount --help" for the complete list of options.
   A useful option is "-o" which can set mount options, in particular:
 
   "-o iocharset=<charset>" for proper encoding of filenames, in case avmount 
   has not autodetected your settings (which should be the default).
   Example :
	avmount -o iocharset=iso-8859-15 /mnt/upnp

   "-o playlists" to render all Audio or Video files as playlists (.m3u or 
   .ram), which contains an URL for the file. In this mode, only the address 
   of the file is exported through avmount ; the actual content is streamed 
   through HTTP when the playlist is accessed by your favorite media player.
   This mode was the only mode possible for avmount before version 0.50.

   "-o search_history=<size>" to set the maximum number of remembered searches
   (see "avmount --help" for the default number). Set to 0 to disable searching
   completely (no "_search" directory will be displayed, even if supported by 
   the connected device).


Known Compatible Devices
------------------------

avmount has been tested successfully with the following UPnP AV 
Devices / Media Servers. Although avmount should work with other 
UPnP AV compliant devices, help is needed to test on additional devices.


* TwonkyVision UPnP Music Server ( http://www.twonkyvision.de )
  versions 2.7.2, 2.9.1, 3.0 and 3.1 for Linux are ok

* GMediaServer on Linux ( http://www.gnu.org/software/gmediaserver/ )
  versions 0.7.0 and 0.9.0 (version 0.8.0 not ok)

* Ahead Nero MediaHome Server ( http://www.nero.com/en/Nero_MediaHome.html )
  included in Nero ultra 6.6

* Microsoft Windows Media Connect 
  ( http://www.microsoft.com/windows/windowsmedia/devices/wmconnect/ )
  requires Windows XP Service Pack 2 (SP2) and .NET FrameWork 1.1. 

* GeeXboX uShare on Linux ( http://ushare.geexbox.org/ )
  versions 0.9.3 to 0.9.7

* AV Media Server in Intel Tools for UPnP Technologies (.NET), 
  ( http://www.intel.com/cd/ids/developer/asmo-na/eng/downloads/upnp/tools/index.htm )
  version 1.0.1768.24080 (Build 1768)

* Nokia N93 phone


This page describes how to share A/V content with some of the above servers :
 http://www.geexbox.org/wiki/index.php/Accessing_to_UPnP_Contents


Troubleshooting
---------------

* if avmount refuses to start, make sure that the FUSE kernel module has
  been loaded before running avmount (try 'modprobe fuse' as root).

* if avmount starts but does not detect your devices and logs the following
  UPnP error -2XX (socket error), make sure that there is a multicast route
  set for the default network interface. This should be done automatically on
  recent distributions, else issue this command (where 'eth0' is the default interface) :
	route add -net 239.0.0.0 netmask 255.0.0.0 dev eth0
  or	route add -net 224.0.0.0 netmask 240.0.0.0 dev eth0

  Another possibility is that libupnp was compiled with IPv6 support but
  it's running on a system without IPv6 support. In that case the solution
  is to either rebuild libupnp without IPv6 support and then rebuild avmount
  or just enable IPv6 support on the kernel and make sure that the module
  is loaded.

* if your UPnP Media Server is behind a closed firewall, it won't be 
  visible in avmount. Therefore, if there is a firewall on the machine
  where you are running your media server, you have to open the necessary
  ports in this firewall. Some media servers can do it automatically 
  (e.g. Windows Media Connect on Windows XP), most of them don't do it 
  (e.g. AV Media Server in Intel Tools).

* in some distributions (e.g. Debian Sarge), avmount (and more generally, 
  FUSE-based file systems) work fine from the command line, but is unbrowsable
  by Nautilus (Gnome) or Konqueror (KDE) : clicking on a directory causes 
  the new browser window to open momentarily, then immediately close, and 
  the directory then disappears from the GUI altogether. 
  The problem is caused by the distro using the older "famd" daemon for file 
  update notification : by default, FUSE doesn't allow root to look at a 
  mounted filesystem (see FUSE documentation). You'll need to add the line :
	 user_allow_other
  to the file "/etc/fuse.conf". Then, you'll need to add the option 
  "-o allow_root" to your mounts.
  Recent distributions (Ubuntu, Mandriva, Fedora, ...) normally don't have 
  this problem : they use the newer "inotify", and doesn't rely on "famd".

* for general problems, try also the FUSE documentation http://fuse.sf.net/ .

* finally, if you still have problems, please report on the avmount issue tracker:
  https://github.com/fernando-rodriguez/avmount/issues


Implementation
--------------

This project uses the following main libraries :

* the FUSE package ( http://fuse.sourceforge.net/ ), 

* the cURL Library ( https://curl.haxx.se/ ),

* the Linux SDK for UPnP Devices (libupnp), initially created by Intel:
  http://upnp.sourceforge.net/.

* the "talloc" library : http://talloc.samba.org/

* for testing only (test_upnp executable) : "readline" library (optionnal)


The UPnP AV Architecture is described here : 
http://www.upnp.org/standardizeddcps/mediaserver.asp


Copyright and Licenses
----------------------

avmount is free software :


* "avmount" is (C) Copyright 2005-2006 Rémi Turboult.
               (C) Copyright 2016 Fernando Rodriguez

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or

NEWS

What is new in avmount 0.8

* First avmount release

* Built-in stream engine

* Unbundled all libraries

* Support for streaming files larger than 2GB

* Multi-home and dynamic network configuration support

* IPv6 support (must be enabled on libupnp and libcurl) *UNTESTED*

* Improved performance and stability

* Fix: memory leaks.

=============================================================================

What is new in djmount 0.71

* fix: filter UPnP devices so that only MediaServers are listed e.g. UPnP 
  Routers are no longer visible (problem reported by Martin Vallevand). 

* fix: correctly handle UPnP devices with embedded sub-devices

* fix: memory corruption when discovering some UPnP devices

* fix: suppress error messages for missing optional elements in device 
  description (URLBase, presentationURL).

* fix: "make check" failures e.g. on Gentoo (fix by Thomas Sattler).

* try to accommodate broken UPnP devices who do not report the MIME type 
  of the Device description document as XML.

* add "text/plain" MIME type to be able to display .txt files

* minor documentation updates

* minor optimisations when analysing XML documents

* some code improvements and more unitary tests

============================================================================

What is new in djmount 0.70

* implement "search" operations (major evolution) : see examples in README
  and details in "search_help.txt" (also available at run-time in "_search/"
  sub-directories).

* reorganise directories to avoid clutter : only multimedia files are directly
  visible. Other items are moved in hidden directories : all metadata XML 
  files are now in sub-directories ".metadata", and debug information 
  (e.g. "status" files) are in "/.debug" directory.

* add some missing MIME types used by GeeXboX uShare

* improve network cache

* improve some error messages

* tested ok with Nokia N93 phone (reported by Sami Kyöstilä)

* code refactoring and miscellaneous code improvements

============================================================================

What is new in djmount 0.53

* allow the 'mmap' operation to succeed on exported files. This helps some
  media players to work with djmount e.g. mpg321 (problem reported by 
  Frank Scholz).
  Note: this works only if using FUSE >= 2.4. Also, this doesn't work with
  some (very rare) media servers who do not tell the size of exported files 
  (i.e. AV Media Server in Intel Tools).

============================================================================

What is new in djmount 0.52

* allow some extra FUSE-related mount options. Mainly : fsname, allow_other, 
  allow_root, nonempty

* adds backward compatibility with FUSE 2.2 (djmount 0.50 initially required 
  FUSE >= 2.4). This is to allow install on Debian Sarge with FUSE stable.

* if using external libupnp, requires at least version 1.3.1

============================================================================

What is new in 0.51

* fix major bug causing empty "browse" on some type of UPnP AV servers,
  for example AV Media Server in Intel Tools for UPnP Technologies
  (reported by Christian Fredrik Kalager Schaller)

* improve djmount freezing when trying to access unreachable Media Servers 
  e.g. behind a closed firewall.

  Note : if your UPnP Media Server is behind a closed firewall, it won't be 
  visible in djmount. Therefore, if there is a firewall on the machine
  where you are running your media server, you have to open the necessary
  ports in this firewall. Some media servers can do it automatically 
  (e.g. Windows Media Connect on Windows XP), most of them don't do it 
  (e.g. AV Media Server in Intel Tools).

* djmount now works ok with AV Media Server in Intel Tools for UPnP 
  Technologies (.NET) (Build 1768)

* tested ok with latest releases of the following Media Servers :
  TwonkyVision UPnP Music Server 3.0, GMediaServer 0.9.0, GeeXboX uShare 0.9.5

* fix configure problem when building with FUSE 2.5.0

* fix warnings when building with gcc4 (tested ok on Ubuntu 5.10, gcc 4.0.2)

* force playlist mode for AV files of unknown sizes (streams, or size 
  not provided by server).

* add matroska and ogg MIME types

* miscellaneous bug fixes

============================================================================

What is new in 0.50

* major change : the content of all files is now directly shared by djmount,
  instead of playlists e.g. MP3 files are directly visible, instead of M3U 
  files. This allow : 
    - seeking A/V files in media players not supporting seeks in HTTP streams,
    - accessing files without playlist formats e.g. images.
  Known limitation : files larger than 2 Gb are not always treated correctly.

  The old djmount behaviour (sharing of playlists) can still be obtained using
  the "-o playlists" mount option (only for A/V files though).

  Note: this change requires using FUSE >= 2.4

* configure: add new --disable-charset option to suppress all character 
  conversion code (i.e. assume everything is in UTF-8)

* fix crash when browsing GeeXboX uShare 0.9.3 (empty directory name)

* fix potential corruption of filenames with international characters

* fix: relative paths to mount points are now working in daemon mode 
  (reported by Andre Landwehr)

* code refactoring and miscellaneous code improvements.


src/djfs.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 2 -*- */
/* $Id: djfs.c 252 2006-08-15 14:08:59Z r3mi $
 *
 * djfs : file system implementation for avmount.
 * This file is part of avmount.
 *
 * (C) Copyright 2016 Fernando Rodriguez
 * (C) Copyright 2005-2006 Rémi Turboult <r3mi@users.sourceforge.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifdef HAVE_CONFIG_H
#	include "../config.h"
#endif

#include "djfs_p.h"
#include "didl_object.h"
#include "file_buffer.h"
#include "media_file.h"
#include "talloc_util.h"
#include "log.h"
#include "content_dir.h"
#include "device_list.h"
#include "xml_util.h"
#include "device.h"

#include "search_help.h"

#include <ctype.h>


/*****************************************************************************

 <mount point> -+- <devName> -+- .status -> ../.debug/<devName>/status
                |             |
                |             +- xxx -+- xxx
                |             |       |- xxx
                |             |       `- _search/
                |             |
                |             +- .metadata/ -+- xxx
                |             |              `- xxx
                |             |
                |             `- _search/ -+- search_capabilities
                |                          |- <search_criteria> -+- ...
                |                          `- ...
                |
                |- <devName>
                |
                `- .debug/

*****************************************************************************/


typedef struct _SearchHistory {

  unsigned int 	serial;
  time_t	time;
  const char*	parent_path; 
  const char*	basename;
  const char*	criteria;
  
} SearchHistory;



/*****************************************************************************
 * BrowseSearchDir
 *****************************************************************************/

static VFS_BrowseStatus
BrowseChildren (DJFS* const self, const char* const sub_path,
		const VFS_Query* const query, void* const tmp_ctx,
		const char* const devName, const DIDLObject* const parent, 
		bool const searchable, const char* const search_criteria,
		ContentDir_Children* const children);

static VFS_BrowseStatus
BrowseSearchDir (DJFS* const self, const char* const sub_path,
		 const VFS_Query* const query, void* const tmp_ctx,
		 const char* const devName, const DIDLObject* const parent, 
		 const char* const criteria_start)
{
  // Keep a pointer to acquired lock, if any
  ithread_mutex_t* lock = NULL;

  BROWSE_BEGIN (sub_path, query) {
    
    FILE_BEGIN ("search_help.txt") {
      static const char* const str = SEARCH_HELP_TXT_STRING;
      FILE_SET_STRING (str, FILE_BUFFER_STRING_EXTERN);
    } FILE_END;

    FILE_BEGIN ("search_capabilities") {
      const char* caps = NULL;
      DEVICE_LIST_CALL_SERVICE (caps, devName, 
				CONTENT_DIR_SERVICE_TYPE,
				ContentDir, GetSearchCapabilities, NULL);
      if (caps && *caps) {
	char* const str = talloc_asprintf (tmp_ctx, "%s\n", caps);
	char* s;
	for (s = str; *s != NUL; s++) {
	  if (*s == ',')
	    *s = '\n';
	}
	FILE_SET_STRING (str, FILE_BUFFER_STRING_STEAL);
      }
    } FILE_END;

    char* const parent_path = talloc_strdup (tmp_ctx, query->path);
    char* pp = parent_path + (BROWSE_PTR - query->path);
    do {
      *pp-- = NUL;
    } while (*pp == '/');

    /*
     * List existing Search criterias
     */
    SearchHistory* h;
    ithread_mutex_lock (&self->search_hist_mutex);
    lock = &self->search_hist_mutex;
    PTR_ARRAY_FOR_EACH_PTR (self->search_hist, h) {
      if (strcmp (parent_path, h->parent_path) == 0) {
	DIR_BEGIN (h->basename) {
	  VFS_SET_TIME (h->time);
	  const char* const full_criteria = (criteria_start ? talloc_asprintf
					     (tmp_ctx, "%s%s)",
					      criteria_start, h->criteria) 
					     : h->criteria);
	  const ContentDir_BrowseResult* res;
	  DEVICE_LIST_CALL_SERVICE (res, devName, 
				    CONTENT_DIR_SERVICE_TYPE,
				    ContentDir, Search,
				    tmp_ctx, parent->id, full_criteria);
	  if (res) {
	    BROWSE_SUB (BrowseChildren (self, BROWSE_PTR, query,
					tmp_ctx, devName, parent,
					true, full_criteria,
					res->children));
	  }
	} DIR_END;
      }
    } PTR_ARRAY_FOR_EACH_PTR_END;
	  
    /*
     * Create new Search criteria
     */
    char* const new_basename = talloc_strdup (tmp_ctx, BROWSE_PTR);
    if (new_basename && *new_basename) {
      bool simplified = true;
      char* t;
      for (t = new_basename; *t != NUL && *t != '/'; t++) {
	if (isspace (*t) || *t == '*' || *t == '"')
	  simplified = false;
      }
      *t = NUL;
      
      char* const new_criteria = 
	(simplified ? talloc_asprintf (tmp_ctx, 
				       "(dc:title contains \"%s\") or "
				       "(dc:creator contains \"%s\") or "
				       "(upnp:artist contains \"%s\") or "
				       "(upnp:album contains \"%s\")",
				       new_basename, new_basename,
				       new_basename, new_basename)
	 : new_basename);
      Log_Printf (LOG_DEBUG, "new search criteria '%s' (inside '%s')",
		  new_criteria, NN(criteria_start));
      const char* const full_criteria = (criteria_start ? talloc_asprintf 
					 (tmp_ctx, "%s%s)", 
					  criteria_start, new_criteria) 
					 : new_criteria);
      const ContentDir_BrowseResult* res;
      DEVICE_LIST_CALL_SERVICE (res, devName, 
				CONTENT_DIR_SERVICE_TYPE,
				ContentDir, Search,
				tmp_ctx, parent->id, full_criteria);
      // Do not create directory on empty result -> "No such file or directory"
      if (res && res->children && 
	  PtrArray_GetSize (res->children->objects) > 0) {
	SearchHistory* h = talloc (self->search_hist, SearchHistory);
	if (h) {
	  *h = (SearchHistory) {
	    .serial      = ++(self->search_hist_serial),
	    .time        = time (NULL),
	    .parent_path = talloc_steal (h, parent_path),
	    .basename	 = talloc_steal (h, new_basename),
	  };
	  h->criteria = ( (new_criteria == new_basename) ? h->basename 
			  : talloc_steal (h, new_criteria) );

	  PtrArray_Append (self->search_hist, h);
	  if (PtrArray_GetSize (self->search_hist) > self->search_hist_size) {
	    SearchHistory* const oldest = 
	      PtrArray_RemoveAt (self->search_hist, 0);
	    talloc_free (oldest);
	  }
	  DIR_BEGIN (new_basename) {
	    VFS_SET_TIME (h->time);
	    BROWSE_SUB (BrowseChildren (self, BROWSE_PTR, query,
					tmp_ctx, devName, parent,
					true, full_criteria,
					res->children));
	  } DIR_END;
	}
      }
    }

  } BROWSE_END;  
  
  // Release any acquired lock
  if (lock)
    ithread_mutex_unlock (lock);
  
  return BROWSE_RESULT;
}


/*****************************************************************************
 * BrowseChildren
 *****************************************************************************/

static VFS_BrowseStatus
BrowseChildren (DJFS* const self, const char* const sub_path,
		const VFS_Query* const query, void* const tmp_ctx,
		const char* const devName, const DIDLObject* const parent, 
		bool const searchable, const char* const search_criteria,
		ContentDir_Children* const children)
{
  // Keep a pointer to acquired lock, if any
  ithread_mutex_t* lock = NULL;
 
  BROWSE_BEGIN(sub_path, query) {
    
    if (children) {
      DIDLObject* o = NULL;               
#if CONTENT_DIR_HAVE_CHILDREN_MUTEX
      ithread_mutex_lock (&children->mutex);
      lock = &children->mutex;
#endif
      PTR_ARRAY_FOR_EACH_PTR (children->objects, o) {
	if (o->is_container) {
	  DIR_BEGIN (o->basename) {
	    const ContentDir_BrowseResult* res;
	    DEVICE_LIST_CALL_SERVICE (res, devName,
				      CONTENT_DIR_SERVICE_TYPE,
				      ContentDir, Browse,
				      tmp_ctx, o->id,
				      CONTENT_DIR_BROWSE_DIRECT_CHILDREN);
	    if (res && res->children) {
	      // Note : if we are already inside a "_search" directory 
	      // ("search_criteria" not NULL), do not allow sub-search 
	      // (might be confusing)
	      BROWSE_SUB (BrowseChildren 
			  (self, BROWSE_PTR, query, tmp_ctx, 
			   devName, o,
			   searchable && (search_criteria == NULL),
			   NULL, res->children));
	    }
	  } DIR_END;
	} else {
	  MediaFile file = { .o = NULL };
	  if (MediaFile_GetPreferred (o, &file)) {
	    off_t const res_size = MediaFile_GetResSize (&file);
	    if ( file.playlist &&
		 ( (self->flags & DJFS_USE_PLAYLISTS) ||
		   res_size < 0) ) {
	      char* name = MediaFile_GetName (tmp_ctx, o, file.playlist);
	      FILE_BEGIN (name) {
		const char* const str = MediaFile_GetPlaylistContent 
		  (&file, tmp_ctx);
		FILE_SET_STRING (str, FILE_BUFFER_STRING_STEAL);
	      } FILE_END;
	    } else {
	      char* name = MediaFile_GetName (tmp_ctx, o, file.extension);
	      FILE_BEGIN (name) {
		FILE_SET_URL (file.uri, res_size);
	      } FILE_END;
	    }
	  }
	}
      } PTR_ARRAY_FOR_EACH_PTR_END;

      if ( (self->flags & DJFS_SHOW_METADATA) && 
	   !PtrArray_IsEmpty(children->objects) ) {
	DIR_BEGIN (".metadata") {
	  PTR_ARRAY_FOR_EACH_PTR (children->objects, o) {
	    char* const name = MediaFile_GetName (tmp_ctx, o, "xml");
	    FILE_BEGIN (name) {
	      const char* const str = talloc_asprintf
		(tmp_ctx, "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n%s",
		 DIDLObject_GetElementString (o, tmp_ctx));
	      FILE_SET_STRING (str, FILE_BUFFER_STRING_STEAL);
	    } FILE_END;
	  } PTR_ARRAY_FOR_EACH_PTR_END;
	} DIR_END;
      }
    } // if children
    
    if (searchable && parent->searchable) {
      // Check if new search directory, or extending a previous search
      if (search_criteria == NULL) {
	DIR_BEGIN ("_search") {
	  BROWSE_SUB (BrowseSearchDir (self, BROWSE_PTR, query,
				       tmp_ctx, devName, parent, NULL));
	} DIR_END;
      } else {
	char criteria_start [strlen (search_criteria) + 9];
	DIR_BEGIN ("_and") {
	  sprintf (criteria_start, "(%s) and (", search_criteria);
	  BROWSE_SUB (BrowseSearchDir (self, BROWSE_PTR, query,
				       tmp_ctx, devName, parent, 
				       criteria_start));
	} DIR_END;
	DIR_BEGIN ("_or") {
	  sprintf (criteria_start, "(%s) or (", search_criteria);
	  BROWSE_SUB (BrowseSearchDir (self, BROWSE_PTR, query,
				       tmp_ctx, devName, parent, 
				       criteria_start));
	} DIR_END;
      }
    } // if searchable
  } BROWSE_END;  
  
  // Release any acquired lock
  if (lock)
    ithread_mutex_unlock (lock);
  
  return BROWSE_RESULT;
}


/*****************************************************************************
 * BrowseRoot
 *****************************************************************************/

static VFS_BrowseStatus
BrowseRoot (VFS* const vfs, const char* const sub_path,
	    const VFS_Query* const query, void* const tmp_ctx)
{
  DJFS* const self = (DJFS*) vfs;
  
  BROWSE_BEGIN(sub_path, query) {
    
    const PtrArray* const names = DeviceList_GetDevicesNames (tmp_ctx);
    
    const char* devName;
    PTR_ARRAY_FOR_EACH_PTR (names, devName) {
      DIR_BEGIN (devName) {
	if (self->flags & DJFS_SHOW_DEBUG) {
	  SYMLINK_BEGIN (".status") {
	    const char* const str = talloc_asprintf (tmp_ctx, 
						     "../%s/%s/status",
						     VFS_DEBUG_DIR_BASENAME,
						     devName);
	    SYMLINK_SET_PATH (str);
	  } SYMLINK_END;
	}
	const ContentDir_BrowseResult* current = NULL;
	DEVICE_LIST_CALL_SERVICE (current, devName, 
				  CONTENT_DIR_SERVICE_TYPE,
				  ContentDir, Browse,
				  tmp_ctx, "0", 
				  CONTENT_DIR_BROWSE_METADATA);
	if (current && current->children) {
	  const DIDLObject* const root =
	    PtrArray_GetHead (current->children->objects);
	  if (root) {
	    DEVICE_LIST_CALL_SERVICE (current, devName, 
				      CONTENT_DIR_SERVICE_TYPE,
				      ContentDir, Browse,
				      tmp_ctx, "0", 
				      CONTENT_DIR_BROWSE_DIRECT_CHILDREN);
	    bool searchable = (self->search_hist != NULL);
	    if (searchable && root->searchable) {
	      // make sure the device is really searchable (some buggy 
	      // servers return "searchable" as "true" in metadata, even
	      // though SearchCapabilities are empty)
	      const char* caps = NULL;
	      DEVICE_LIST_CALL_SERVICE (caps, devName, 
					CONTENT_DIR_SERVICE_TYPE,
					ContentDir, GetSearchCapabilities, 
					NULL);
	      searchable = (searchable && caps && *caps);
	    }
	    if (current && current->children) {
	      BROWSE_SUB (BrowseChildren (self, BROWSE_PTR, query, tmp_ctx, 
					  devName, root, searchable,
					  NULL, current->children));
	    }
	  }
	}
      } DIR_END; // devName
    } PTR_ARRAY_FOR_EACH_PTR_END;
    
  } BROWSE_END;
  
  return BROWSE_RESULT;
}


/*****************************************************************************
 * BrowseDebug
 *****************************************************************************/
#ifdef DEBUG
static VFS_BrowseStatus
BrowseDebug (VFS* const vfs, const char* const sub_path,
	    const VFS_Query* const query, void* const tmp_ctx)
{
  DJFS* const self = (DJFS*) vfs;
  
  // Keep a pointer to acquired lock, if any
  ithread_mutex_t* lock = NULL;

  BROWSE_BEGIN(sub_path, query) {
        
    FILE_BEGIN("versions") {
      static const char* const str = PACKAGE " " VERSION "\nlibupnp "
#ifdef UPNP_VERSION_STRING
	UPNP_VERSION_STRING
#else
	"bundled"
#endif
	"\nFUSE " STRINGIFY (FUSE_MAJOR_VERSION) "." 
	STRINGIFY (FUSE_MINOR_VERSION) "\n";
      FILE_SET_STRING (str, FILE_BUFFER_STRING_EXTERN);
    } FILE_END;

    // Status of each device
    const PtrArray* const names = DeviceList_GetDevicesNames (tmp_ctx);
    const char* devName;
    PTR_ARRAY_FOR_EACH_PTR (names, devName) {
      DIR_BEGIN (devName) {
	FILE_BEGIN ("status") {
	  const char* const str = 
	    DeviceList_GetDeviceStatusString (tmp_ctx, devName, true);
	  FILE_SET_STRING (str, FILE_BUFFER_STRING_STEAL);
	} FILE_END;

	FILE_BEGIN ("device_description.xml") {
	  // Must use FILE_SET_STRING here, instead of FILE_SET_URL with 
	  // the URL of the description document, because the VFS interface
	  // would causes a HTTP-RANGE request for the document, which is 
	  // not allowed for the Device description document (see UPnP Device 
	  // Architecture 1.0).
	  const char* str;
	  DEVICE_LIST_CALL_DEVICE (str, devName, GetDescDocTextCopy, tmp_ctx);
	  FILE_SET_STRING (str, FILE_BUFFER_STRING_STEAL);
	} FILE_END;
      } DIR_END;
    } PTR_ARRAY_FOR_EACH_PTR_END;

    // List historized Search requests
    if (self->search_hist) {
      DIR_BEGIN ("search_history") {
	SearchHistory* h;
	ithread_mutex_lock (&self->search_hist_mutex);
	lock = &self->search_hist_mutex;
	PTR_ARRAY_FOR_EACH_PTR (self->search_hist, h) {
	  char link_name [33];
	  sprintf (link_name, "%d", h->serial);
	  SYMLINK_BEGIN (link_name) {
	    const char* const str = talloc_asprintf 
	      (tmp_ctx, "../..%s/%s", h->parent_path, h->basename);
	    SYMLINK_SET_PATH (str);
	    VFS_SET_TIME (h->time);
	  } SYMLINK_END;
	} PTR_ARRAY_FOR_EACH_PTR_END;
      } DIR_END;
    }

    // Call superclass' method
    VFS_BrowseFunction func = CLASS_METHOD (VFS, browse_debug);
    if (func) {
      BROWSE_SUB (func (vfs, BROWSE_PTR, query, tmp_ctx));
    }
    
  } BROWSE_END;
    
  // Release any acquired lock
  if (lock)
    ithread_mutex_unlock (lock);
  
  return BROWSE_RESULT;
}
#endif

/******************************************************************************
 * finalize
 *
 * Description: 
 *	DJFS destructor
 *
 *****************************************************************************/

static void
finalize (Object* obj)
{
  DJFS* const self = (DJFS*) obj;

  if (self && self->search_hist) {
    ithread_mutex_destroy (&self->search_hist_mutex);
  }

  // Other "talloc'ed" fields will be deleted automatically : 
  // nothing to do 
}


/*****************************************************************************
 * OBJECT_INIT_CLASS
 *****************************************************************************/

// Initialize class methods
static void
init_class (DJFS_Class* const isa)
{
  CLASS_BASE_CAST(isa)->finalize      = finalize;
  CLASS_SUPER_CAST(isa)->browse_root  = BrowseRoot;
#ifdef DEBUG
  CLASS_SUPER_CAST(isa)->browse_debug = BrowseDebug;
#endif
}

OBJECT_INIT_CLASS(DJFS, VFS, init_class);


/*****************************************************************************
 * DJFS_Create
 *****************************************************************************/
DJFS*
DJFS_Create (void* talloc_context, DJFS_Flags flags, 
	     size_t search_history_size)
{
  OBJECT_SUPER_CONSTRUCT (DJFS, VFS_Create, 
			  talloc_context, (flags & DJFS_SHOW_DEBUG));
  if (self) {
    self->flags              = flags;
    self->search_hist_size   = search_history_size;
    self->search_hist_serial = 0;

    if (search_history_size > 0) {
      self->search_hist = PtrArray_Create (self);
      if (self->search_hist == NULL) {
	talloc_free (self);
	self = NULL;
      } else {
	ithread_mutexattr_t attr;
	ithread_mutexattr_init (&attr);
	ithread_mutexattr_setkind_np (&attr, ITHREAD_MUTEX_RECURSIVE_NP);
	ithread_mutex_init (&self->search_hist_mutex, &attr);
	ithread_mutexattr_destroy (&attr);
      }
    }
  }
  return self;
}



